ethereumj-core/src/main/java/org/ethereum/config/Initializer.java

```diff
@@ -54,4 +63,92 @@ public Object postProcessBeforeInitialization(Object bean, String beanName) thro
     public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
         return bean;
     }
+
+    /**
+     * We need to persist the DB version, so after core upgrade we can either reset older incompatible db
+     * or make a warning and let the user reset DB manually.
+     * Database version is stored in ${database}/version.properties
+     * Logic will assume that database has version 1 if file with version is absent.
+     */
+    public static class IncompatibleDatabaseHandler {
+
+        public enum Behavior {
+            EXIT, RESET, IGNORE
+        }
+
+        public void process(SystemProperties config) {
+            final File versionFile = new File(config.databaseDir() + "/version.properties");
+            final Behavior behavior = Behavior.valueOf(
+                    config.getProperty("database.incompatibleDatabaseBehavior", Behavior.EXIT.toString()).toUpperCase());
+
+
+            // Detect database version
+            final Integer expectedVersion = config.databaseVersion();
+            if (isDatabaseDirectoryExists(config)) {
+                final Integer actualVersionRaw = getDatabaseVersion(versionFile);
+                final boolean isVersionFileNotFound = actualVersionRaw.equals(-1);
+                final Integer actualVersion = isVersionFileNotFound ? 1 : actualVersionRaw;
+
+                if (actualVersionRaw.equals(-1)) {
+                    putDatabaseVersion(versionFile, actualVersion);
+                }
+
+                if (actualVersion.equals(expectedVersion) || (isVersionFileNotFound && expectedVersion.equals(1))) {
+                    logger.info("Database directory location: '{}', version: {}", config.databaseDir(), actualVersion);
+                } else {
+                    logger.warn("Detected incompatible database version. Detected:{}, required:{}", actualVersion, expectedVersion);
+                    if (behavior == Behavior.EXIT) {
+                        System.err.println("Please remove database directory manually or set `database.incompatibleDatabaseBehavior` to `RESET`");
+                        System.err.println("Database directory location is " + config.databaseDir());
+                        throw new Error("Incompatible database version " + actualVersion + ". Please remove database " +
+                                "directory manually or set `database.incompatibleDatabaseBehavior` to `RESET`");
+                    } else if (behavior == Behavior.RESET) {
+                        FileUtil.recursiveDelete(config.databaseDir());
```

Comment:
Can you please also check the return value here and throw appropriate exception. If the directory couldn't be deleted I got the following messy exception later:
```
Caused by: java.lang.RuntimeException: Can't initialize database
	at org.ethereum.datasource.LevelDbDataSource.init(LevelDbDataSource.java:94)
	at org.ethereum.db.RepositoryImpl.init(RepositoryImpl.java:109)
	at org.ethereum.db.RepositoryImpl.<init>(RepositoryImpl.java:86)
	at org.ethereum.config.CommonConfig.repository(CommonConfig.java:58)
	at org.ethereum.config.CommonConfig$$EnhancerBySpringCGLIB$$6d19a2ca.CGLIB$repository$1(<generated>)
	at org.ethereum.config.CommonConfig$$EnhancerBySpringCGLIB$$6d19a2ca$$FastClassBySpringCGLIB$$4f30a1b2.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:228)
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:318)
	at org.ethereum.config.CommonConfig$$EnhancerBySpringCGLIB$$6d19a2ca.repository(<generated>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:162)
	... 41 more
Caused by: java.nio.file.AccessDeniedException: D:\ws\ethereumj\work\database
	at sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:83)
	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:97)
	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:102)
	at sun.nio.fs.WindowsFileSystemProvider.createDirectory(WindowsFileSystemProvider.java:504)
	at java.nio.file.Files.createDirectory(Files.java:674)
	at java.nio.file.Files.createAndCheckIsDirectory(Files.java:781)
	at java.nio.file.Files.createDirectories(Files.java:767)
	at org.ethereum.datasource.LevelDbDataSource.init(LevelDbDataSource.java:77)
```